### PR TITLE
Zero order

### PR DESCRIPTION
**Pull request by hcorg**

### CODE REVIEW

File: Signals/Domain/Services/Implementation/SignalsDomainService.cs [hunk 1]
```diff
@@ -104,7 +106,13 @@ private IEnumerable<Datum<T>> FillMissingData<T>(Signal signal, TimeEnumerator t
             var missingValuePolicy = GetMissingValuePolicy(signal)
                 as MissingValuePolicy.MissingValuePolicy<T>;
 
-            return missingValuePolicy.FillMissingData(timeEnumerator, readData);
+            var olderData = missingValuePolicy.OlderDataSamplesCountNeeded > 0 
+                ? this.signalsDataRepository.GetDataOlderThan<T>(signal, timeEnumerator.FromIncludedUtc, 1)
```
Comment (bartoszkp) [inline Signals/Domain/Services/Implementation/SignalsDomainService.cs]: Shouldn't the last parameter be `OlderDataSamplesCountNeeded`?

File: Signals/Domain/MissingValuePolicy/MissingValuePolicy.cs [hunk 2]
```diff
@@ -27,9 +27,11 @@ public abstract class MissingValuePolicy<T> : MissingValuePolicyBase
         [NHibernateIgnore]
         public override Type NativeDataType { get { return typeof(T); } }
 
-        public abstract IEnumerable<Datum<T>> FillMissingData(TimeEnumerator timeEnumerator, IEnumerable<Datum<T>> readData);
-        /*NoneQuality,
-        SpecificValue,
-        Interpolation*/
+        [NHibernateIgnore]
+        public virtual int OlderDataSamplesCountNeeded { get { return 0; } }
```
Comment (bartoszkp) [inline Signals/Domain/MissingValuePolicy/MissingValuePolicy.cs]: Should be `OlderDataSampleCountNeeded` (count already carries plural semantics)

File: Signals/Domain/MissingValuePolicy/ZeroOrderMissingValuePolicy.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+            var nextValueTs = readEnumerator.MoveNext() ? readEnumerator.Current.Timestamp : timeEnumerator.ToExcludedUtcUtc;
+
+            foreach (var ts in timeEnumerator)
+            {
+                if (ts == nextValueTs)
+                {
+                    lastQuality = readEnumerator.Current.Quality;
+                    lastValue = readEnumerator.Current.Value;
+                    nextValueTs = readEnumerator.MoveNext() ? readEnumerator.Current.Timestamp : timeEnumerator.ToExcludedUtcUtc;
+                }
```
Comment (bartoszkp) [inline Signals/Domain/MissingValuePolicy/ZeroOrderMissingValuePolicy.cs:29]: \n

File: SignalsIntegrationTests/SignalsIntegrationTests/Infrastructure/MissingValuePolicyTestsBase.cs [hunk 4]
```diff
@@ -29,12 +29,17 @@ protected void GivenASignal(Granularity granularity)
...
+            GivenDatums();
         }
 
         protected void GivenSingleDatum(Datum<int> datum)
         {
-            client.SetData(signalId, new[] { datum }.ToDto<Dto.Datum[]>());
+            GivenDatums(datum);
+        }
+
+        protected void GivenDatums(params Datum<int>[] datums)
```
Comment (bartoszkp) [inline SignalsIntegrationTests/SignalsIntegrationTests/Infrastructure/MissingValuePolicyTestsBase.cs]: Datums -> Data

File: SignalsIntegrationTests/SignalsIntegrationTests/ZeroOrderPolicyTests.cs [hunk 5]
```diff
@@ -0,0 +1,146 @@
+﻿using Domain;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using SignalsIntegrationTests.Infrastructure;
+using System;
```
Comment (bartoszkp) [inline SignalsIntegrationTests/SignalsIntegrationTests/ZeroOrderPolicyTests.cs]: Go to Tools -> Options -> C# -> Advanced (IIRC) and check the option for system usings to be first.

Comment (hcorg) [inline SignalsIntegrationTests/SignalsIntegrationTests/ZeroOrderPolicyTests.cs]: done.
still, it treats Microsoft.VisualStudio.TestTools.UnitTesting as non system using :(

File: Signals/Domain/Services/Implementation/SignalsDomainService.cs (hunk 1 above)
Comment (hcorg) [inline Signals/Domain/Services/Implementation/SignalsDomainService.cs]: thx for catching it up. (already thinking about higher order interpolations. which I will not do. which makes this code unnecessary... :D )

**Comment by hcorg** [comment]
update pushed

**Pull request merged**